presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -358,7 +361,52 @@ else if (firstDependencies.stream().allMatch(right::canResolve) && secondDepende
         return new RelationPlan(root, analysis.getScope(node), outputSymbols);
     }
 
-    private static boolean isEqualComparisonExpression(Expression conjunct)
+    private Optional<Unnest> getUnnest(Relation relation)
+    {
+        if (relation instanceof AliasedRelation) {
+            return getUnnest(((AliasedRelation) relation).getRelation());
+        }
+        if (relation instanceof Unnest) {
+            return Optional.of((Unnest) relation);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<Lateral> getLateral(Relation relation)
+    {
+        if (relation instanceof AliasedRelation) {
+            return getLateral(((AliasedRelation) relation).getRelation());
+        }
+        if (relation instanceof Lateral) {
+            return Optional.of((Lateral) relation);
+        }
+        return Optional.empty();
+    }
+
+    private RelationPlan planLateralJoin(Node joinNode, RelationPlan leftPlan, Lateral lateral)
+    {
+        RelationPlan rightPlan = process(lateral.getQuery(), null);
+        PlanBuilder leftPlanBuilder = initializePlanBuilder(leftPlan);
+        PlanBuilder planBuilder = subqueryPlanner.appendApplyNode(
```

Comment:
I don't think lateral join should be handled by `ApplyNode`. Currently `ApplyNode` has a well defined semantics which can be described like: take row from input, execute subquery with the input row as correlation, pass input row with appended subquery results. If we allow `ApplyNode` to be transformed into cross join with non-scalar subqueries then `ApplyNode` semantics will differ depending if subquery is scalar or not (e.g: `ApplyNode` will replicate input rows or not). This can lead to subtle bugs, where non-scalar subquery `ApplyNodes` will be transformed to `CrossJoins`, while currently they will be left in plan and query will fail.

Additionally, there are two types of lateral join: `LEFT JOIN LATERAL` and `INNER JOIN LATERAL`. PostgreSQL mentions that `LEFT JOIN LATERAL` are particularly useful. They use it in their blog example: http://blog.heapanalytics.com/postgresqls-powerful-new-join-type-lateral/. If lateral join is planned with `ApplyNode`, then you can support only `INNER JOIN LATERAL` and adding `LEFT JOIN LATERAL` will not fit into the code nicely.

Given that, I think there are two other appropriate approaches for lateral join in planner:
1. Add `LateralJoinNode`: This way one can one easily add support for both  `LEFT LATERAL` and `CROSS LATERAL` joins. There are already multiple join nodes in planner: `SemiJoinNode`, `IndexJoinNode`, `JoinNode`. Therefore I think lateral join deserves it's own node too.
2. Plan lateral join with `ApplyNode` and `Unnest`. One can rewrite lateral join into plan nodes:
```
              UnnestNode
              - output_symbols: sym_x, unnest_sym_y
              - unnest_symbols: unnest_sym_y -> sym_y
                          |
                          |
                  ApplyNode
                  - output_symbols: sym_x, sym_y
                  - subquery_projections: sym_y -> array(sub_sym_y)
                 /                                  \
               /                                     \
             /                                        \
       Input node                                Subquery node
       - output_symbols: sym_x                   - output_symbols: sub_sym_y
```
In 2) approach you could add support for `LEFT LATERAL` by adding support for `LEFT UNNEST`. `ApplyNode` projcetion `sym_y -> array(sub_sym_y)` should never be executed. There should be optimizer that recognizes `UnnestNode` above `ApplyNode` with such subquery projection that would transform those nodes directly into left or cross join. This is fragile though since other optimizers might decouple `UnnestNode` from `ApplyNode` by projection pushdown for instance.

Given the complexity of 2) I strongly favor approach 1). What do you think @kokosing ?